### ISSUE TITLE
./toolchain.py create <name> <path_to_main> ... makes wrong YourApp folder

### ISSUE THREAD

**Issue by ssolari** [kivy/kivy-ios#344]
### system
OS: MacOS mojave
Python: 3.7.1
Kivy: 1.10.1
Kivy-ios: master (as of 1/18/19)

### issue
I used to be able to have a path to creating Xcode projects, but now starting with a fresh download of kivy-ios the command ```./toolchain.py create samplekivy path_to_main```  creates ```samplekivy-ios/samplekivy.xcodeproj``` that uses some default YourApp folder.    None of my package files are found and inlcuded in the YourApp folder?    As I said, I wasn't having any problems about 2 weeks ago now I can't even make a basic hello world app.

<img width="219" alt="screen shot 2019-01-18 at 11 28 06 am" src="https://user-images.githubusercontent.com/1182784/51408849-dea28700-1b14-11e9-886e-19957e69c747.png">

**Comment by ssolari** [comment]
@tshirtman just wanted to quickly get a guage ... is this an actual issue [i.e. are others seeing this?]. It seems like this is a show stopper, or am doing something unbelievable stupid?  As I said I had things working but now cant figure out why it doesn't pickup the main.py and subpackages.    I just completely reinstalled kivy-ios and see the same thing. 

i have a basic folder structure like:
  * myapp
    * myapp
      * __ init__.py
      * main.py
      * subpackage1
        *  __ init__.py
      * subpackage2
        *  __ init__.py
    * buildscripts.py
    * other stuff

I clone the kivy-ios folder run 
```./toolchain.py build kivy python3```   
That works, then run 
```./toolchain.py create testapp ../myapp/myapp```

And I get the default YourApp folder.    Also, it would be nice if kivy-ios threw an error or something if it was going to use the default YourApp folder and couldn't find a main.py from the user.    I feel like that should just not be allowed.

**Comment by hackalog** [comment]
What happens if you do the same thing, but with python 2?

**Comment by ssolari** [comment]
All our code is python3 so it won't work in python 2, but tried to setup just getting a button or something to display with python2.   I couldn't build getting the following error from pyobjus
```
Downloading https://github.com/kivy/pyobjus/archive/master.zip
Traceback (most recent call last):
  File "./toolchain.py", line 1486, in <module>
    ToolchainCL()
  File "./toolchain.py", line 1221, in __init__
    getattr(self, args.command)()
  File "./toolchain.py", line 1262, in build
    build_recipes(args.recipe, ctx)
  File "./toolchain.py", line 1099, in build_recipes
    recipe.execute()
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 671, in execute
    self.download()
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 62, in _cache_execution
    f(self, *args, **kwargs)
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 706, in download
    self.download_file(self.url.format(version=self.version), fn)
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 454, in download_file
    urlretrieve(url, filename, report_hook)
  File "/Users/sorensolari/anaconda/envs/kivy2env/lib/python2.7/site-packages/future/backports/urllib/request.py", line 1759, in retrieve
    block = fp.read(bs)
  File "/Users/sorensolari/anaconda/envs/kivy2env/lib/python2.7/site-packages/future/backports/http/client.py", line 516, in read
    return bytes(super(HTTPResponse, self).read(amt))
  File "/Users/sorensolari/anaconda/envs/kivy2env/lib/python2.7/site-packages/future/backports/http/client.py", line 545, in readinto
    return self._readinto_chunked(b)
  File "/Users/sorensolari/anaconda/envs/kivy2env/lib/python2.7/site-packages/future/backports/http/client.py", line 654, in _readinto_chunked
    n = self._safe_readinto(temp_mvb)
  File "/Users/sorensolari/anaconda/envs/kivy2env/lib/python2.7/site-packages/future/backports/http/client.py", line 701, in _safe_readinto
    n = self.fp.readinto(mvb)
AttributeError: '_fileobject' object has no attribute 'readinto'
```

For pythonn 3 and python 2, I am using a conda environment to run the initial builds.  
for the python 2 I ran
```conda env create -n kivy2env -f environment.yml```
### environment.yml
```
channels:
  - https://conda.anaconda.org/conda-forge
dependencies:
  - python=2.7
  - pip:
    - kivy==1.10.1
    - pbxproj==2.5.1
```

then
```
source activate kivy2env
./toolchain.py build kivy python2
```

**Comment by hackalog** [comment]
You can work around the `_fileobject` error with https://github.com/kivy/kivy-ios/pull/332

(p.s. @tito is there any objection to merging that one? it clearly doesn't hit just me)

**Comment by hackalog** [comment]
Yes, i’m mainly just wondering if you see the same bug is in the python2
path. (I’m pretty sure it is not there, as I use that toolchain daily). If
you can reproduce it on a hello world (especially if it turns out to be
python3 only), that's perfect.

I will try to reproduce this soon, but I likely won’t get to it until the
weekend. Sorry!

**Comment by ssolari** [comment]
@hackalog ok thanks.   I'll try to get a python2 version up and running tomorrow to see if I can reproduce.    Any suggestions on where to start to debug?  It seems like it would be in the cookiecutter somewhere, but I've tried to go down that path and I'm a bit lost of what that cookiecutter is doing. 

I had a nice build before some of the merges that happened.   When I pulled the new master this started happening.   However, I have also tried to hard reset to an old commit on master and I still see the same bug, so I'm wondering if anything could have happenend with paths?   But I am using clean conda environments as well as pulling and rebuilding kivy-ios from scratch so I don't know what paths would be picked up.    Either way, the main.py should be found from the path passed at the command line to the toolchain.

The good (and challenge for these moments) is that I'm completely committed to helping figure out a clean path for a python3 build for an app we are building and plan to deploy on ios and android.

**Comment by hackalog** [comment]
I'd be *really* surprised if any of the january changes caused the change. (I would have guessed the bigger changes in October/November). Were you running a kivy-ios master from december at any point? Did things work then?

If so, I'd try before and after this change: https://github.com/kivy/kivy-ios/commit/271e235927ab39cf9f08e0d4f0736414a03392e8#diff-17d1f5bba3152b61f1d788c2da85971f  since that changed the cookiecutter project

**Comment by ssolari** [comment]
I was definitely running something from december.   I'm going to go back even further as you suggested.   Working on the urlretrieve issue for building python 2.    After the error above, I can't seem to clean the pyobjus and now getting the following error:
```
Download pyobjus
Traceback (most recent call last):
  File "./toolchain.py", line 1486, in <module>
    ToolchainCL()
  File "./toolchain.py", line 1221, in __init__
    getattr(self, args.command)()
  File "./toolchain.py", line 1262, in build
    build_recipes(args.recipe, ctx)
  File "./toolchain.py", line 1099, in build_recipes
    recipe.execute()
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 671, in execute
    self.download()
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 62, in _cache_execution
    f(self, *args, **kwargs)
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 707, in download
    status = self.get_archive_rootdir(self.archive_fn)
  File "/Users/sorensolari/Documents/Simpa/code_stuff/kivy-ios/toolchain.py", line 501, in get_archive_rootdir
    with zipfile.ZipFile(filename) as zf:
  File "/Users/sorensolari/anaconda/envs/kivy2env/lib/python2.7/zipfile.py", line 770, in __init__
    self._RealGetContents()
  File "/Users/sorensolari/anaconda/envs/kivy2env/lib/python2.7/zipfile.py", line 811, in _RealGetContents
    raise BadZipfile, "File is not a zip file"
zipfile.BadZipfile: File is not a zip file
```

I had patched pyobjus and kivy recipes in my own forks in the python3 version.   May go back to that to see if I can make it work.    If I go back to a version of master far enough back then I know its some other issue.

**Comment by ssolari** [comment]
I have reverted to what was a working version of a fork i had from kivy-ios and i still see the same thing.   I am at a loss.    What paths could have changed?    As I said this happened because I wanted to try to keep pace with the master branch of kivy-ios.   Specifically so I could try to help the ongoing development and uncover bugs/fixes to help.   But I did not anticipate completely destroying things like this (which seem so trivially simple). 

Where is the path to the main.py file picked up and then how are the packages within that directory searched?   I need to fix this.

**Comment by hackalog** [comment]
You have a bogus (0-sized) zipfile lying around from when your download failed (before applying #332). Delete that and the zipfile error will go away. (I forget offhand where that cached zip file is stored, and I'm not at a computer. the build dir probably?)

I do, in fact, run kivy-ios using a conda environment (which was some of the motivation for fixing https://github.com/kivy/kivy-ios/pull/331, so that whatever pip/python you have enabled in a conda environment will work correctly)

**Comment by ssolari** [comment]
@hackalog thanks.   Can you help me figure out where exactly when (or in what part of the code) the project files are copied to the "Sources" directory?    I'm looking through all the cookiecutter code trying to figure out where exactly that happens.     I figure if I can locate that moment when those files are copied then I can see why its not happening.    Any insight as to where in the code that happens?

**Comment by ssolari** [comment]
Same thing is happening with python2.   I have to assume something changed in my specific environment if no one else is experiencing this.   However, I think it would be good (and its necessary on my side) to figure out the root cause as this has been a nightmare and would be for anyone else.

My debug goal is to find out exactly where the package files are copied to Sources.   Obviously that line is not being executed and from there I can just backtrack out why its not being executed.   I am amazed how hard it is to figure out where this is happening.   

@hackalog ANY help here would be enormously appreciated.   I have been looking through generate.py in cookiecutter but can not figure it out.   I have no idea where the 'project_dir' in the context (which is the path to my package) is ever used. 

The only place it looks like that path is used is in line 221 in project.pbxproj:
```shellScript = "rsync -av --delete \"{{ cookiecutter.project_dir }}\"/ \"$PROJECT_DIR\"/YourApp";```
I'm not that familiar with xcode or cookiecutter, but this line seems to be only excecuted by xcode?   So then I don't know when the app package files would/should be copied when using ```./toolchain.py create or update```?

FYI. my fix to the other issues was to simply replace the urlretrieve line with:
```
        import requests
        response = requests.get(url, stream=True)
        handle = open(filename, "wb")
        for chunk in response.iter_content(chunk_size=512):
            if chunk:  # filter out keep-alive new chunks
                handle.write(chunk)
        handle.close()
```
This has seemed to be far more robust than the urlretrieve (which I actually had to fix for another issue too).  Maybe think about using requests?  It would simplify a lot of code.   I think even including request module in kivy to run the UrlRequest might be beneficial.  While an extra dependency, we use it in our kivy code because it more obvious what is happening and faster.

**Comment by hackalog** [comment]
Did you apply the fix in https://github.com/kivy/kivy-ios/pull/332 ? It
does use requests as a fallback. Requests, however, does not support ftp in
a URL, (and we have at least one recipe that still requires ftp) so is not
a replacement as is. See the issue for justification and discussion.

I will have some time to look into the python 3 process this weekend. The
rsync command is run when you press “play” in xcode.

Did you try running that command manually to see if there is an error
occurring in your environment?

**Comment by ssolari** [comment]
The original issue is that I don't have any of my original files in the YourApp directory in Xcode.  Effectively my kivy app is ignored completely.    

I don't think this is a python2 or python3 issue.    The problem is that a ```./toolchain.py create <name> <path_to_my_app>```  is somehow ignoring <path_to_my_app>?   Or it doesn't see it.   The toolchain has to copy at least the main.py in the <path_to_my_app> into the "YourApp" folder for my app to run in Xcode, but that copy never happens.    I can't figure out why not.

**Comment by hackalog** [comment]
You asked where the code is supposed to be copied.

As you already noticed,  rsync is supposed to copy your code from the project dir to `YourApp`: 
"rsync -av --delete \"{{ cookiecutter.project_dir }}\"/ \"$PROJECT_DIR\"/YourApp"

This only happens AFTER you press play in XCode though. If you have tried it within xcode, and it still doesn't work, I suggested that you run the rsync command manually to ensure no errors were occurring.

The python2/python3 conversation was to determine if it was your setup, or the python3 path of kivy-ios. It appears that it is likely your setup. 

The reference to #332 was because you proposed an alternate solution to `urlretrieve` failing, and asked about using `requests` instead, which the fix in #332 is already doing.

**Comment by ssolari** [comment]
@hackalog thank you for bearing with me.   I can't believe just not pressing play in xcode was the problem.   I must have gotten accustomed to doing updates and seeing the code in the YourApp folder as it was already there.   Then when I did a clean start from the beginning I never pressed build because I saw the default YourApp folder with everything inside it.   I can't believe how many hours I spent chasing non existent ghosts. 

However this whole thing and experience with kivy-ios still makes me think that there may be a more streamlined approach to deploying a kivy app.    I have enough experience that if I am having so many problems that it would be very hard for someone to come in without a lot of experience and get something up and running.   That should not be the case, kivy is just too amazing to keep it for only those that have a lot of experience (and determination to work through this build process).     Part of my comments in #345 .   As you use conda I am guessing you must have some opinions.   I'll let that issue be a place for any discussion, but kivy-ios was started many years ago, back when honestly the tools like conda were still somewhat new or underdeveloped.    Again thanks for your help even if to help me see my own failure.   I'm closing this issue as it is solved.
